examples/hamiltonian_representation.py

```diff
@@ -0,0 +1,322 @@
+
+    Args:
+        hamiltonian_polynomial_list: the list of Hamiltonians, typically built by calling
+            build_hamiltonian_from_boolean().
+        qubits: The list of qubits corresponding to the variables.
+        theta: A single float scaling the rotations.
+        ladder_target: Whether to use convention of figure 7a or 7b.
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+    combined = sum(hamiltonian_polynomial_list, HamiltonianPolynomial.O())
+
+    circuit = cirq.Circuit()
+
+    # Here we follow improvements of [4] cancelling out the CNOTs. The first step is to order by
+    # Gray code so that as few as possible gates are changed.
+    sorted_hs = sorted(
+        list(combined.hamiltonians.keys()), key=functools.cmp_to_key(_gray_code_comparator)
+    )
+
+    def _apply_cnots(ph: Tuple[int, ...], ch: Tuple[int, ...]):
+        # This function applies in sequence the CNOTs from ph and then ch. However, given that the
+        # h are sorted in Gray ordering and that some cancel each other, we can reduce the number
+        # of gates. See [4] for more details.
+
+        cnots: List[Tuple[int, int]] = []
+
+        if ladder_target:
+            cnots.extend((ph[i], ph[i + 1]) for i in reversed(range(len(ph) - 1)))
+            cnots.extend((ch[i], ch[i + 1]) for i in range(len(ch) - 1))
+        else:
+            cnots.extend((ph[i], ph[-1]) for i in range(len(ph) - 1))
+            cnots.extend((ch[i], ch[-1]) for i in range(len(ch) - 1))
+
+        while True:
+            # As per equations 9 and 10, if all the targets (resp. controls) are the same, the
+            # cnots commute. Further, if the control (resp. targets) are the same, the cnots can be
+            # simplified away.
+            found_simplification = False
+            for x, y in [(0, 1), (1, 0)]:
+                i = 0
+                for j in range(1, len(cnots)):
+                    if cnots[i][x] != cnots[j][x]:
+                        # The targets (resp. control) don't match, so we reset the search.
+                        i = j
+                        continue
+                    if cnots[i][y] == cnots[j][y]:
+                        # The controls (resp. targets) are the same, so we can simplify away.
+                        cnots = [cnots[k] for k in range(len(cnots)) if k != i and k != j]
+                        found_simplification = True
+                        break
+
+            if found_simplification:
+                continue
+
+            # Here we apply the simplification of equation 11. Note that by flipping the control
+            # and target qubits, we have an equally valid identity:
+            # CNOT(i, j) @ CNOT(j, k) == CNOT(j, k) @ CNOT(i, k) @ CNOT(i, j)
+            # CNOT(j, i) @ CNOT(k, j) == CNOT(k, j) @ CNOT(k, i) @ CNOT(j, i)
+            #
+            # Which are represented by (x, y) = (0, 1) and (x, y) = (1, 0), respectively.
```

Comment:
I read the original paper again and it sounds very vague to me as well. They said it can help to improve but doesn't give any systematic approach. The approach you used is greedy as well. I don't know better solutions.  Your key idea is repeatedly to try the above three CNOTs to two CNOTs, right?

The complexity of code is out of my original imagination...  (I was thinking about some simple improvements..)

Reply:
Ack. Maybe we leave it as is and wait to see whether they publish more?